### ISSUE TITLE
Relative links not working in asciidoc render

### ISSUE THREAD

**Issue by deparcus** [go-gitea/gitea#26203]
### Description

I activated asciidoc file render based on https://docs.gitea.com/next/administration/external-renderers:
* install asciidoctor via `apt install asciidoctor`
* add external render in `app.ini`
```
[markup.asciidoc]
ENABLED = true
FILE_EXTENSIONS = .adoc,.asciidoc
RENDER_COMMAND = "asciidoctor -s -a showtitle --out-file=- -"
; Input is not a standard input but a file
IS_INPUT_FILE = false
```

Then I created a branch in one of my repositories and migrated Markdown files to adoc files.
In the main `*.adoc` file I use relative links to link to other `*.adoc` files.
Therefore I followed the [asciidoctor documentation](https://docs.asciidoctor.org/asciidoc/latest/macros/link-macro/#link-to-a-relative-file).

Example file `docs/README.adoc`:
```
= Network documentation

== Index

* link:general/[General]
** link:general/netplan.drawio.svg[Netplan]
** link:general/Patchpanels.adoc[Patchpanels]
** link:general/static-IPs.adoc[Static IPs]
```

I also tried other relative links:
```
= Network documentation

== Index

* link:./general/[General]
** link:./general/netplan.drawio.svg[Netplan]
** link:./general/Patchpanels.adoc[Patchpanels]
** link:./general/static-IPs.adoc[Static IPs]
```

In VSCode the links are working and I can navigate to the files in subdirectories by clicking the link in the main file.

The link is interpreted as it starts from the root of the branch. So `/docs/` is missing. But as the `README.adoc` is in the folder `docs`, the links should be interpreted as starting from there.

This is shown in the screenshot below:
![gitea_relative_adoc_links_not_working_correctly](https://github.com/go-gitea/gitea/assets/128035935/e8a1c03e-2b06-4f51-b839-865b8bdba172)

I discovered that issue also in another repository, where I first tried it.

When I use Markdown (currently my master branch), everything works as is should.

Example file `doc/README.md`:
```
# Network documentation

## Index

* [General](./general/)
  * [Netplan](./general/netplan.drawio.svg)
  * [Patchpanels](./general/Patchpanels.md)
  * [Static IPs](./general/static-IPs.md)
```

Working links, see screenshot:
![gitea_relative_md_links_working_correctly](https://github.com/go-gitea/gitea/assets/128035935/97c69389-419d-4e44-ab14-2681d8102fb1)

Am I doing something wrong?
Or is there maybe an issue with `asciidoctor` or the `asciidoctor` command in `app.ini`?

Would be great if anybody could help me.

### Gitea Version

1.20.1

### Can you reproduce the bug on the Gitea demo site?

No

### Screenshots

![gitea_relative_adoc_links_not_working_correctly](https://github.com/go-gitea/gitea/assets/128035935/e8a1c03e-2b06-4f51-b839-865b8bdba172)
![gitea_relative_md_links_working_correctly](https://github.com/go-gitea/gitea/assets/128035935/97c69389-419d-4e44-ab14-2681d8102fb1)

### Git Version

2.30.2

### Operating System

Raspbian/Debian 11.7

### How are you running Gitea?

Local install with binary and systemd service.

### Database

MySQL

**Comment by Mo-Gul** [comment]
I have exactly the same issue on two different Gitea's.

## tl;dr of the above is

- When the `docs/README.adoc` is active/selected (the address bar shows `<repository>/src/branch/main/docs/README.adoc`), the link destination works as expected, i.e. are `<repository>/src/branch/main/docs/<link_dest>`.
- When "only" the `docs` folder is selected (the address bar shows `<repository>/src/branch/main/docs`), the link destination is `<repository>/src/branch/main/<link_dest>`.
- When being in the root folder (the address bar shows `<repository>`), the link destination is `<repository>/<link_dest>`.

With

- `<repository>` for this repository would be `https://github.com/go-gitea/gitea`,
- `<link_dest>` would be e.g. `link.adoc`, and
- the link is written as `link:<link_dest>.adoc[]` in the `docs/README.adoc` file.

**Comment by Mo-Gul** [comment]
Because GitHub must have had the same issue (I guess), I found github/markup#1039 which might be a good start so find out, how it was solved there.

Hopefully then also `xref`s can be handled or -- depending on `safe-mode` -- `include` directives are converted automatically to links (as can be seen e.g. at https://github.com/opendevise/asciidoc-samples/blob/main/runtime.adoc)

Maybe @mojavelinux could as least give a hint on how to solve the issue here (as well) ...

**Comment by mojavelinux** [comment]
Please don't @ me in projects I'm not involved with. You are welcome to ask in the Asciidoctor project chat at https://chat.asciidoctor.org.

**Comment by Mo-Gul** [comment]
@deparcus, do you want to ask for help in the asciidoctor chat since you started the issue? If not, please let me know and I'll try to do it myself.

**Comment by jtiee** [comment]
The bug is that Gitea does not add a trailing slash in links to repos/folders.

Consider this situation:

The user views `https://<gitea-host>/<user>/<repository>`. Gitea automatically renders the `README` file (`README.adoc` in this case, but any format would suffer the same problem).

Relative links in the auto-rendered `README` might look like `docs/index.adoc`. When you click such a link, the browser combines the current URL context with the relative link. The combination uses the existing "folder" (path segment) in the current URL as the context.

What "folder" is that? `https://<gitea-host>/<user>` because there is no trailing slash after `<repository>`. So, the browser then attempts to fetch `https://<gitea-host>/<user>/docs/index.adoc`, which results in a 404.

If Gitea adds the trailing slash to repo/folder links, then when the user clicks the `docs/index.adoc` link, the browser would fetch `https://<gitea-host>/<user>/<repository>/docs/index.adoc`, which is correct.

You can manually add a trailing slash (and reload the page) to any repo/folder view and click a relative link in an auto-generated `README` file to confirm the correct behavior.

I used `gitea-1.20.3-darwin-10.12-amd64` for testing.

**Comment by Mo-Gul** [comment]
Seems you are right and the "same" problem exists for markdown files as well. See #18592.

**Comment by Mo-Gul** [comment]
I can confirm that with Gitea v1.19.0 a from a README.md in the docs folder work to another MD file in the docs folder works regardless of how the README.md is shown.

Unfortunately, I couldn't find out why so far. But maybe this has something to do with the [Slugify rules for Gitea](https://markdown-all-in-one.github.io/docs/specs/slugify/gitea.html).

@deparcus, could you setup some dummy Gitea servers with previous versions so we could drill down since *when* the links are working? Maybe then it is easier to find out exactly why/how this is working for markdown files and thus there hopefully is a good start for a asciidoctor solution.

**Comment by Mo-Gul** [comment]
Not really, but I guess I would start with v1.11 and then do a [binary search](https://en.wikipedia.org/wiki/Binary_search_algorithm) when it isn't working there up to <v1.19 where it is working. Maybe it also a good idea to start with [v1.16.1](https://github.com/go-gitea/gitea/releases/tag/v1.16.1) which release date is before the above mentioned bug report from me.

**Comment by Mo-Gul** [comment]
I don't have much of a clue about docker, so I trust whatever you are doing. But maybe a bit simpler test is 
[markdown_test-main.zip](https://github.com/go-gitea/gitea/files/12903500/markdown_test-main.zip) where there is only one `README.md` in the `docs` folder and the link should work when you are in the base repository directory.

Good luck finding the Gitea version where the link isn't working (properly) any more!!

**Comment by Mo-Gul** [comment]
Since I currently can't setup (easily) other Gitea's myself and I trust your results, I agree that this maybe is a dead end.
I also already tried to find the related code parts where links are handled but I most likely I didn't succeed because I don't know anything about the (programming) language.
So hopefully you have more luck than I do ...

**Comment by Mo-Gul** [comment]
That is the obvious way, but I have not much hope here, because up to now nobody else than jtiee commented here. I'd be happy to proven wrong.
Unfortunately the same seems to be true for the asciidoctor chat ... :cry:

**Comment by jtiee** [comment]
I've updated my PoC so that I have AsciiDoc and Markdown files with the same content:

```
$ ls -la
total 16
drwxr-xr-x   6 eedwards  staff  192  2 Nov 09:38 ./
drwxr-xr-x   3 eedwards  staff   96 25 Aug 18:12 ../
drwxr-xr-x  13 eedwards  staff  416  2 Nov 09:42 .git/
-rw-r--r--   1 eedwards  staff   69 25 Aug 18:15 README.adoc
-rw-r--r--   1 eedwards  staff   67  2 Nov 09:38 README.md
drwxr-xr-x  12 eedwards  staff  384  2 Nov 09:41 docs/
$ ls -la docs
total 80
drwxr-xr-x  12 eedwards  staff  384  2 Nov 09:41 ./
drwxr-xr-x   6 eedwards  staff  192  2 Nov 09:38 ../
-rw-r--r--   1 eedwards  staff   79  2 Nov 09:38 READ.md
-rw-r--r--   1 eedwards  staff   81 25 Aug 18:29 README.adoc
-rw-r--r--   1 eedwards  staff   30 25 Aug 18:14 development.adoc
-rw-r--r--   1 eedwards  staff   33  2 Nov 09:39 development.md
-rw-r--r--   1 eedwards  staff  139 25 Aug 18:15 index.adoc
-rw-r--r--   1 eedwards  staff  129  2 Nov 09:39 index.md
-rw-r--r--   1 eedwards  staff   31 25 Aug 18:14 install.adoc
-rw-r--r--   1 eedwards  staff   34  2 Nov 09:39 install.md
-rw-r--r--   1 eedwards  staff   26 25 Aug 18:14 support.adoc
-rw-r--r--   1 eedwards  staff   29  2 Nov 09:39 support.md
```

I used `docs/READ.md` instead of `docs/README.md` to avoid auto-rendering the Markdown file.

Then I compared the links rendered by Gitea in `docs/index.adoc` and `docs/index.md`. First, here is the markup for the links:

`docs/index.adoc`:
```
* link:install.adoc[Installation]
```

`docs/index.md`:
```
- [Installation](install.md)
```

As you can see, both implement relative links. When Gitea renders these:

`docs/index.adoc`:
```
<a href="install.adoc" rel="nofollow">Installation</a>
```

`docs/index.md`:
```
<a href="/eskwayrd/test/src/branch/main/docs/install.md" rel="nofollow">Installation</a>
```

As you can see, the rendered Asciidoc files have the relative links as defined in the file. Whereas the Markdown files have an absolute link from the document root. That's the problem.

I don't know if Gitea is making the rendered Markdown links absolute or whether the Markdown rendered used by Gitea is doing that. But nothing is making the AsciiDoc links absolute.

Either Gitea needs to add slashes after folder names, or it needs to make relative links absolute in rendered AsciiDoc files.

**Comment by lunny** [comment]
There are two envs to help with your asciidoc command. `GITEA_PREFIX_SRC` and `GITEA_PREFIX_RAW` could help you to make the links absolute.


About how to use, please see https://docs.gitea.com/next/administration/config-cheat-sheet?_highlight=gitea_prefix_src#markup-markup

**Comment by Mo-Gul** [comment]
@lunny, I have stumbled over these variables at different places so far, e.g. also at https://github.com/go-gitea/gitea/issues/6359#issuecomment-1059958752, but unfortunately I couldn't find out how _exactly_ to use these (with a concrete example).

I just gave it some tries with

```
RENDER_COMMAND="asciidoctor --base-dir=$GITEA_PREFIX_SRC --no-header-footer --attribute=showtitle --out-file=- -"
RENDER_COMMAND="asciidoctor --base-dir=GITEA_PREFIX_SRC --no-header-footer --attribute=showtitle --out-file=- -"
RENDER_COMMAND="asciidoctor --attribute docdir=$GITEA_PREFIX_SRC --no-header-footer --attribute=showtitle --out-file=- -"
```

but unfortunately none of them worked. What am I doing wrong it these environment variables (really) should be able to solve the problem?

**Comment by lunny** [comment]
Since recently, markup render system has many refactors, I need to confirm whether this issue have been fixed.

**Comment by wxiaoguang** [comment]
New Gitea should have fixed various relative path problems. The real fix is https://github.com/go-gitea/gitea/issues/34114 (in 1.24)

Some related details:

1. NEED_POST_PROCESS=true (defaults to true in 1.24, changed name in 1.25) will make Gitea resolve the relative URLs 
2. Make external render use environment variables like `GITEA_PREFIX_SRC`


Please try Gitea 1.24 stable or 1.25 nightly, feel free to reopen if there is still any problem
